<a href="https://colab.research.google.com/github/SonakshiA/Apriori-Algorithm-for-Market-Basket-Analysis/blob/main/Apriori_Algorithm_for_Market_Basket_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Importing Libraries**

In [ ]:
import numpy as np
import pandas as pd
import mlxtend
from mlxtend.frequent_patterns import apriori, association_rules

**Reading the Data**

In [ ]:
df = pd.read_csv('Retail_Data.csv')


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


**Displaying the Data**

In [ ]:
df.head()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Trans_ID Product1   Product2 Product3
0         1    Bread     Butter    Dairy
1         2   Nachos     Butter    Dairy
2         3    Juice        Jam      Egg
3         4    Juice        Jam      Egg
4         5    Juice  Vegetable    Salad

**One-Hot Encoding**

In [4]:
#Create dummy variables
df1 = pd.get_dummies(df) #The get_dummies function in pandas is used to convert categorical variables into dummy/indicator variables.
df2 = df1.iloc[:,1:] #all rows and all columns starting from the second column
df2.head()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Product1_Bread  Product1_Fruits  Product1_Juice  Product1_Nachos  \
0            True            False           False            False   
1           False            False           False             True   
2           False            False            True            False   
3           False            False            True            False   
4           False            False            True            False   

   Product2_Butter  Product2_Jam  Product2_Salsa  Product2_Vegetable  \
0             True         False           False               False   
1             True         False           False               False   
2            False          True           False               False   
3            False          True           False               False   
4            False         False           False                True   

   Product3_Dairy  Product3_Egg  Product3_Salad  
0            True         False           False  
1            True         False           False  
2           False          True           False  
3           False          True           False  
4           False         False            True

**Support:** Refers to the number of transactions of an item/ set of items with respect to all the transactions.

In [10]:
#Rules having atleast 5% support
frequent_items = apriori(df2,min_support=0.05, use_colnames=True)
frequent_items

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


support                                           itemsets
0    0.3510                                   (Product1_Bread)
1    0.1825                                  (Product1_Fruits)
2    0.2465                                   (Product1_Juice)
3    0.2200                                  (Product1_Nachos)
4    0.2465                                  (Product2_Butter)
5    0.3410                                     (Product2_Jam)
6    0.1965                                   (Product2_Salsa)
7    0.2160                               (Product2_Vegetable)
8    0.2030                                   (Product3_Dairy)
9    0.5760                                     (Product3_Egg)
10   0.2210                                   (Product3_Salad)
11   0.2145                  (Product1_Bread, Product2_Butter)
12   0.1280                     (Product2_Jam, Product1_Bread)
13   0.1005                   (Product3_Dairy, Product1_Bread)
14   0.2505                     (Product3_Egg, Product1_Bread)
15   0.1705              (Product1_Fruits, Product2_Vegetable)
16   0.1825                  (Product1_Fruits, Product3_Salad)
17   0.2010                     (Product2_Jam, Product1_Juice)
18   0.2080                     (Product1_Juice, Product3_Egg)
19   0.1880                  (Product2_Salsa, Product1_Nachos)
20   0.1025                  (Product3_Dairy, Product1_Nachos)
21   0.1175                    (Product3_Egg, Product1_Nachos)
22   0.0520                  (Product3_Dairy, Product2_Butter)
23   0.1945                    (Product3_Egg, Product2_Butter)
24   0.0720                     (Product3_Dairy, Product2_Jam)
25   0.2570                       (Product2_Jam, Product3_Egg)
26   0.0790                   (Product3_Dairy, Product2_Salsa)
27   0.1175                     (Product2_Salsa, Product3_Egg)
28   0.2090               (Product3_Salad, Product2_Vegetable)
29   0.1945    (Product3_Egg, Product1_Bread, Product2_Butter)
30   0.0720     (Product3_Dairy, Product2_Jam, Product1_Bread)
31   0.0560       (Product2_Jam, Product3_Egg, Product1_Bread)
32   0.1705  (Product1_Fruits, Product2_Vegetable, Product3...
33   0.2010       (Product1_Juice, Product2_Jam, Product3_Egg)
34   0.0705  (Product3_Dairy, Product1_Nachos, Product2_Salsa)
35   0.1175    (Product2_Salsa, Product3_Egg, Product1_Nachos)

**Confidence:** It is the probability of buying product B given the consumer has bought product A

**Lift:** Its value indicates the level of association between the antecedent and consequents.
* Lift<1 suggest a negative association
* Lift=1 suggests no association
* List>1 suggests a strong association

Lift is the ratio of confidence to support

In [12]:
#Rules with minimum lift of 1
rules = association_rules(frequent_items,metric="lift",min_threshold=1)
rules

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents                        consequents  \
0                    (Product1_Bread)                  (Product2_Butter)   
1                   (Product2_Butter)                   (Product1_Bread)   
2                      (Product2_Jam)                   (Product1_Bread)   
3                    (Product1_Bread)                     (Product2_Jam)   
4                    (Product3_Dairy)                   (Product1_Bread)   
..                                ...                                ...   
65  (Product2_Salsa, Product1_Nachos)                     (Product3_Egg)   
66    (Product3_Egg, Product1_Nachos)                   (Product2_Salsa)   
67                   (Product2_Salsa)    (Product3_Egg, Product1_Nachos)   
68                     (Product3_Egg)  (Product2_Salsa, Product1_Nachos)   
69                  (Product1_Nachos)     (Product2_Salsa, Product3_Egg)   

    antecedent support  consequent support  support  confidence      lift  \
0               0.3510              0.2465   0.2145    0.611111  2.479153   
1               0.2465              0.3510   0.2145    0.870183  2.479153   
2               0.3410              0.3510   0.1280    0.375367  1.069420   
3               0.3510              0.3410   0.1280    0.364672  1.069420   
4               0.2030              0.3510   0.1005    0.495074  1.410467   
..                 ...                 ...      ...         ...       ...   
65              0.1880              0.5760   0.1175    0.625000  1.085069   
66              0.1175              0.1965   0.1175    1.000000  5.089059   
67              0.1965              0.1175   0.1175    0.597964  5.089059   
68              0.5760              0.1880   0.1175    0.203993  1.085069   
69              0.2200              0.1175   0.1175    0.534091  4.545455   

    leverage  conviction  zhangs_metric  
0   0.127978    1.937571       0.919316  
1   0.127978    4.999328       0.791820  
2   0.008309    1.039009       0.098504  
3   0.008309    1.037260       0.100022  
4   0.029247    1.285337       0.365138  
..       ...         ...            ...  
65  0.009212    1.130667       0.096552  
66  0.094411         inf       0.910482  
67  0.094411    2.195079       1.000000  
68  0.009212    1.020092       0.184906  
69  0.091650    1.894146       1.000000  

[70 rows x 10 columns]

In [16]:
#filter out rules having support>=2, confidence>=0.6 and support>=0.2
rules[(rules['lift']>=2) & (rules['support']>=0.2) & (rules['confidence']>=0.6)]

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents                   consequents  \
0                 (Product1_Bread)             (Product2_Butter)   
1                (Product2_Butter)              (Product1_Bread)   
13                (Product1_Juice)                (Product2_Jam)   
32                (Product3_Salad)          (Product2_Vegetable)   
33            (Product2_Vegetable)              (Product3_Salad)   
53  (Product1_Juice, Product3_Egg)                (Product2_Jam)   
54    (Product2_Jam, Product3_Egg)              (Product1_Juice)   
55                (Product1_Juice)  (Product2_Jam, Product3_Egg)   

    antecedent support  consequent support  support  confidence      lift  \
0               0.3510              0.2465   0.2145    0.611111  2.479153   
1               0.2465              0.3510   0.2145    0.870183  2.479153   
13              0.2465              0.3410   0.2010    0.815416  2.391249   
32              0.2210              0.2160   0.2090    0.945701  4.378247   
33              0.2160              0.2210   0.2090    0.967593  4.378247   
53              0.2080              0.3410   0.2010    0.966346  2.833860   
54              0.2570              0.2465   0.2010    0.782101  3.172824   
55              0.2465              0.2570   0.2010    0.815416  3.172824   

    leverage  conviction  zhangs_metric  
0   0.127978    1.937571       0.919316  
1   0.127978    4.999328       0.791820  
13  0.116944    3.570187       0.772141  
32  0.161264   14.438667       0.990498  
33  0.161264   24.037714       0.984181  
53  0.130072   19.581714       0.817076  
54  0.137650    3.458027       0.921700  
55  0.137650    4.025264       0.908857

**Conclusion:**

We extrapolate the first row:
1.   21% of all transactions are bread and butter.
2.   61% consumers buy butter given they buy bread.
3. Since lift>1, there is a strong association between bread and butter, retailers should arrange them next to each other for more sales and higher profits.



Some other metrics:

**Leverage:** Leverage measures the difference between the observed frequency of co-occurrence of two items (X and Y) in transactions and the frequency that would be expected if X and Y were independent of each other. In other words, it quantifies how much the occurrence of X and Y together deviates from what would be expected by chance.

*Leverage(X->Y) = Support(X and Y) - Support(X)Support(Y)*

* A positive leverage value indicates that the occurrence of X and Y together is more frequent than what would be expected by chance, suggesting a positive association.

* A leverage value of 0 suggests that the co-occurrence of X and Y is as expected under independence, meaning there is no association.

* A negative leverage value indicates that the occurrence of X and Y together is less frequent than expected, suggesting a negative association or avoidance.


